---
title: "R_assignment"
author: "Mudith Ekanayake"
date: "3/18/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{r cars}
summary(cars)
```

## Including Plots

You can also embed plots, for example:

```{r pressure, echo=FALSE}
plot(pressure)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.


# BCB546 - UNIX Assignment - Mudith Ekanayake

## Part I

## Data Inspection

### Attributes of fang_et_al_genotypes.txt

```{r}
library(tidyverse)
library(tidyr)
```

To load the fang_et_al_genotypes.txt data file into R
```{r}
fang_data = read_tsv("fang_et_al_genotypes.txt")
```

To get the file size
```{r}
file.size("fang_et_al_genotypes.txt")
```

To get all the file info
```{r}
file.info("fang_et_al_genotypes.txt", extra_cols = TRUE)
```

To compactly display the internal structure of the R object
```{r}
str(fang_data)
```

To get an idea about the data frame by viewing the first and last few rows
```{r}
head(fang_data)
tail(fang_data)
```

To get the dimensions of the data frame
```{r}
dim(fang_data)
```

To get the number of rows in the data frame
```{r}
nrow(fang_data)
```

To get the number of columns in the data frame
```{r}
ncol(fang_data)
```

To get the structure of the data frame by previewing data in the columns
```{r}
str(fang_data)
```

To view the column names
```{r}
names(fang_data)
```

To see the class of all the columns
```{r}
sapply(fang_data, class)
```

By inspecting this file I learned that:

File size: 11051939 bytes
Dimension of the dataframe: 2782 x 986
Number of rows: 2782
Number of columns: 986

### Attributes of snp_position.txt

To load the snp_position.txt data file into R
```{r}
snp_data = read_tsv("snp_position.txt")
```

To get the file size
```{r}
file.size("snp_position.txt")
```

To get all the file info
```{r}
file.info("snp_position.txt", extra_cols = TRUE)
```

To compactly display the internal structure of the R object
```{r}
str(snp_data)
```

To get an idea about the data frame by viewing the first and last few rows
```{r}
head(snp_data)
tail(snp_data)
```

To get the dimensions of the data frame
```{r}
dim(snp_data)
```

To get the number of rows in the data frame
```{r}
nrow(snp_data)
```

To get the number of columns in the data frame
```{r}
ncol(snp_data)
```

To get the structure of the data frame by previewing data in the columns
```{r}
str(snp_data)
```

To view the column names
```{r}
names(snp_data)
```

To see the class of all the columns
```{r}
sapply(snp_data, class)
```

By inspecting this file I learned that:

File size: 82763 bytes
Dimension of the dataframe: 983 x 15
Number of rows: 983
Number of columns: 15

## Data Processing

```{r}
snp_data <- snp_data[c(1,3,4,2,5:15)]
```
snp_data data frame was formatted such that the first column is "SNP_ID", the second column is "Chromosome", the third column is "Position".

### For maize

```{r}
maize <- fang_data %>% filter(Group=="ZMMIL"|Group=="ZMMLR"|Group=="ZMMMR")
```

```{r}
maize <- as.data.frame(t(maize), stringsAsFactors = FALSE)
```

```{r}
SNP_ID <- rownames(maize)
rownames(maize) <- NULL
```

```{r}
maize <- cbind(SNP_ID, maize, stringsAsFactors = FALSE)
```
cbind r function stands for column binding and it used to combine vectors matrices or data frames by columns. It splits matrix columns in data frame arguments and "stringsAsFactors = FALSE" prevents converting character columns to factors.

```{r}
names(maize) <-  c("SNP_ID",maize[1,-1])
maize <- maize[-c(1,2,3), ]
```


```{r}
merged_maize <- merge(snp_data, maize, by="SNP_ID")
```


```{r}
dir.create("maize_data")
```
Create directory for storing generated files for maize


```{r}
for(i in c(1:10)){
  maize_data <- merged_maize %>% filter(Chromosome==i) %>% mutate(Position_new=as.numeric(Position)) %>% arrange(Position_new)
  maize_data$Position_new <- NULL
  write.csv(maize_data, paste0("maize_chr_asc",i,".csv"), row.names = FALSE)
}
```


```{r}
for(i in c(1:10)){
  maize_data <- merged_maize %>% filter(Chromosome==i)%>% mutate(Position_new=as.numeric(Position)) %>% arrange(-Position_new)
  maize_data$Position_new <- NULL
  maize_data[maize_data == "?/?"]<-"-/-"
  write.csv(maize_data, paste0("maize_chr_dsc",i,".csv"), row.names = FALSE)
}
```

### For Teosinte

```{r}
teosinte <- fang_data %>% filter(Group=="ZMPJA"|Group=="ZMPIL"|Group=="ZMPBA")
```

```{r}
teosinte <- as.data.frame(t(teosinte), stringsAsFactors = FALSE)
```

```{r}
SNP_ID <- rownames(teosinte)
rownames(teosinte) <- NULL
```

```{r}
teosinte <- cbind(SNP_ID, teosinte, stringsAsFactors = FALSE)
```
cbind r function stands for column binding and it used to combine vectors matrices or data frames by columns. It splits matrix columns in data frame arguments and "stringsAsFactors = FALSE" prevents converting character columns to factors.

```{r}
names(teosinte) <-  c("SNP_ID", teosinte[1,-1])
teosinte <- teosinte[-c(1,2,3), ]
```


```{r}
merged_teosinte <- merge(snp_data, teosinte, by="SNP_ID")
```


```{r}
dir.create("teosinte_data")
```
Create directory for storing generated files for teosinte


```{r}
for(i in c(1:10)){
  teosinte_data <- merged_teosinte %>% filter(Chromosome==i) %>% mutate(Position_new=as.numeric(Position)) %>% arrange(Position_new)
  teosinte_data$Position_new <- NULL
  write.csv(teosinte_data, paste0("teosinte_chr_asc",i,".csv"), row.names = FALSE)
}
```


```{r}
for(i in c(1:10)){
  teosinte_data <- merged_teosinte %>% filter(Chromosome==i)%>% mutate(Position_new=as.numeric(Position)) %>% arrange(-Position_new)
  teosinte_data$Position_new <- NULL
  teosinte_data[teosinte_data == "?/?"]<-"-/-"
  write.csv(teosinte_data, paste0("teosinte_chr_dsc",i,".csv"), row.names = FALSE)
}
```

## Part II

## SNPs per chromosome

### Plotting the total number of SNPs in the dataset on each chromosome.

is.numeric function checks whether its argument is numerical or not and is.na function checks whether there are missing values. "!" symbol reverses the function of is.na.
```{r}
snp_data_n <- snp_data[!is.na(as.numeric(snp_data$Chromosome)),]
```

geom_bar() function was used to plot the bar chart. scale_x_discrete() function was used to set the values for discrete x aesthetic. labs() function was used to label the x and y axis of the graph. Bar chart was colored according to the "Chromosome".
```{r}
ggplot(data = snp_data_n) + geom_bar(mapping = aes(x = Chromosome, fill = Chromosome)) + scale_x_discrete(limit=c(1:10)) + labs(x = "Chromosome", y = "Number of SNPs")
```

### Plotting the distribution of SNPs on chromosomes.

reshaping the original data using the pivot_longer() function in the tidyr package.Sample_ID, JG_OTU and Group columns were selected. "names_to" specifies the name of the column as "SNP_ID" and the column was created from the data stored in the column names of "fang_data" data frame. "values_to" specifies the name of the column as "NT" and the column was created from the data stored in cell values.
```{r}
fang_pivot <- fang_data %>% pivot_longer(!c(Sample_ID, JG_OTU, Group), names_to="SNP_ID", values_to= "NT")
```

fang_pivot data frame was merged by SNP_ID.
```{r}
merged_fang_Pivot <- merge(fang_pivot, snp_data, by="SNP_ID")
```

```{r}
merged_fang_Pivot_n <- merged_fang_Pivot[!is.na(as.numeric(merged_fang_Pivot$Chromosome)),]
```

Bar chart was colored and shaded according to the "Group".
```{r}
ggplot(data = merged_fang_Pivot_n) + geom_bar(mapping = aes(as.numeric(Chromosome), fill=Group)) + scale_x_discrete(limit=c(1:10)) + labs(x = "Chromosome", y="Distribution of SNPs")
```

## Missing data and amount of heterozygosity

Creating a new column named "homo_or_hetero" and indicating all the sites as "Heterozygous".
```{r}
merged_fang_Pivot$homo_or_hetero <- "Heterozygous"
```

Check for the missing data and replace the sites with "Missing Data" in the "homo_or_hetero" column.
```{r}
merged_fang_Pivot$homo_or_hetero[merged_fang_Pivot$NT == "?/?"] <- "Missing Data"
```

Check for the homozygous sites and replace the sites with "Homozygous" in the "homo_or_hetero" column.
```{r}
merged_fang_Pivot$homo_or_hetero[merged_fang_Pivot$NT %in% c("A/A", "C/C", "G/G", "T/T")] <- "Homozygous"
```

Proportion of homozygous and heterozygous sites as well as missing data in each sample was plotted using ggplot and height of the individual bars were normalized using ggplot's "position adjustment" option. Graph was labeled using labs() function.
```{r}
ggplot(data = merged_fang_Pivot) + geom_bar(mapping=aes(x = Sample_ID, fill = homo_or_hetero), position = "fill") + labs(x = "Sample_ID", y="Proportion")
```

Proportion of homozygous and heterozygous sites as well as missing data in each Group was plotted using ggplot and height of the individual bars were normalized using ggplot's "position adjustment" option. Graph was labeled using labs() function.
```{r}
ggplot(data = merged_fang_Pivot) + geom_bar(mapping = aes(x = Group, fill = homo_or_hetero), position = "fill") + labs(x = "Group", y="Proportion") + ggtitle("")
```


```{r}
ggplot(data = merged_fang_Pivot) + geom_bar(mapping = aes(x = Sample_ID, fill = gene), position = "fill") + labs(x = "Gene", y="Proportion")
```